### PR TITLE
fix(diagnostic): vim.diagnostic.reset errors if buffer was wiped out

### LINKED ISSUE #16391: vim.diagnostic.reset errors if buffer has been wiped out
### Neovim version (nvim -v)

NVIM v0.6.0-dev+607-ga2749482d

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

macOS 12.0.1

### Terminal name/version

iTerm2 Build 3.5.0beta2

### $TERM environment variable

xterm-256color / tmux-256color

### Installation

Build from source

### How to reproduce the issue

Create a Lua file with the following content:

```lua
local ns = vim.api.nvim_create_namespace("test")
local bufnr = vim.fn.bufadd("my-file")
vim.fn.bufload(bufnr)
vim.diagnostic.set(ns, bufnr, { lnum = 0, end_lnum = 0, col = 0, end_col = 0 })

vim.cmd("bwipeout " .. bufnr)

vim.diagnostic.reset(ns)
```

Run the code using `:source` or any other method. 

### Expected behavior

The code runs without errors. 

### Actual behavior

The following error is thrown:

```
Error detected while processing /Users/jose/test.lua:
E5113: Error while calling lua chunk: Vim:E158: Invalid buffer name: 2
stack traceback:
        [C]: in function 'sign_unplace'
        ...rs/jose/.local/share/nvim/runtime/lua/vim/diagnostic.lua:877: in function 'hide'
        ...rs/jose/.local/share/nvim/runtime/lua/vim/diagnostic.lua:1068: in function 'hide'
        ...rs/jose/.local/share/nvim/runtime/lua/vim/diagnostic.lua:1342: in function 'reset'
        test.lua:8: in main chunk
```

### PR DESCRIPTION
**Pull request by smolck**
Should close #16391.

### CODE REVIEW

File: runtime/lua/vim/diagnostic.lua [hunk 1]
```diff
@@ -1049,17 +1049,15 @@ function M.hide(namespace, bufnr)
...
-  end
-
   local buffers = bufnr and {get_bufnr(bufnr)} or vim.tbl_keys(diagnostic_cache)
   for _, iter_bufnr in ipairs(buffers) do
-    local namespaces = namespace and {namespace} or vim.tbl_keys(diagnostic_cache[iter_bufnr])
-    for _, iter_namespace in ipairs(namespaces) do
-      for _, handler in pairs(M.handlers) do
-        if handler.hide then
-          handler.hide(iter_namespace, iter_bufnr)
+    if vim.api.nvim_buf_is_valid(bufnr) then
```
Comment (gpanders) [inline runtime/lua/vim/diagnostic.lua:1054]:
```suggestion
    if vim.api.nvim_buf_is_valid(iter_bufnr) then
```

**Review by gpanders — COMMENTED** [review]
Can you add a test as well? The repro from the initial issue is a good starting point, and let’s also test that `hide()` as well as `reset()` don’t error when the buffer is invalid.

File: runtime/lua/vim/diagnostic.lua [hunk 2]
```diff
@@ -60,7 +60,24 @@ local bufnr_and_namespace_cacher_mt = {
 }
 
 local diagnostic_cleanup = setmetatable({}, bufnr_and_namespace_cacher_mt)
-local diagnostic_cache = setmetatable({}, bufnr_and_namespace_cacher_mt)
+local diagnostic_cache = setmetatable({}, {
+  __index = function(t, bufnr)
```
Comment (smolck) [inline runtime/lua/vim/diagnostic.lua:64]: So this is weird; you'd think `__newindex` would be the way to go here, but it never seems to be called, either here or with `bufnr_and_namespace_cacher_mt`. Maybe `__index` doing the `rawset(t, bufnr, {})` part has something to do with it? Not sure, but it seems weird.

Also this `__index` is basically exactly the same as the one in `bufnr_and_namespace_cacher_mt`, any thoughts on reducing the repetition here, or is it not that big of a deal?

File: runtime/lua/vim/diagnostic.lua [hunk 3]
```diff
@@ -60,7 +60,24 @@ local bufnr_and_namespace_cacher_mt = {
...
+    if not bufnr or bufnr == 0 then
+      bufnr = vim.api.nvim_get_current_buf()
+    end
+
+    if rawget(t, bufnr) == nil then
```
Comment (gpanders) [inline runtime/lua/vim/diagnostic.lua:69]: I don’t think the `rawget` is necessary as `__index` is only called if the key does not exist.

File: runtime/lua/vim/diagnostic.lua [hunk 4]
```diff
@@ -59,8 +59,22 @@ local bufnr_and_namespace_cacher_mt = {
   end,
 }
 
-local diagnostic_cleanup = setmetatable({}, bufnr_and_namespace_cacher_mt)
-local diagnostic_cache = setmetatable({}, bufnr_and_namespace_cacher_mt)
+local diagnostic_cache = setmetatable({}, {
+  __index = function(t, bufnr)
+    if not bufnr or bufnr == 0 then
```
Comment (gpanders) [inline runtime/lua/vim/diagnostic.lua:64]: Is it even possible for `bufnr` to be nil here? If not, let’s remove the useless `not bufnr` check

File: runtime/lua/vim/diagnostic.lua [hunk 5]
```diff
@@ -59,8 +59,22 @@ local bufnr_and_namespace_cacher_mt = {
...
+      bufnr = vim.api.nvim_get_current_buf()
+    end
+
+    vim.api.nvim_buf_attach(bufnr, false, {
+      on_detach = function(_, b)
```
Comment (gpanders) [inline runtime/lua/vim/diagnostic.lua:69]: I don’t think we need to capture the `b` argument here
```suggestion
      on_detach = function()
```

**Review by gpanders — CHANGES_REQUESTED** [review]
One other change request: can you remove the `clear_diagnostic_cache` function and just replace the remaining usages of it with `diagnostic_cache[bufnr][namespace] = nil`? That’s all it does anyway, and there’s no reason to have the extra level of indirection

File: runtime/lua/vim/diagnostic.lua (hunk 4 above)
Comment (smolck) [reply runtime/lua/vim/diagnostic.lua:64]: Tbh I don't actually know?

**Review by gpanders — APPROVED** [review]

**Comment by gpanders** [comment]
Just need to squash and write a commit message detailing the changes

**Comment by clason** [comment]
Just a heads-up: this broke `vim.diagnostic.goto_next()` (at least with LSP).

**Comment by gpanders** [comment]
Can you send me a repro? It seems to still work okay for me with clangd.

**Comment by mjlbach** [comment]
Call `lua vim.diagnostic.get(0)` first

**Comment by gpanders** [comment]
#16405

**Pull request merged**